extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -552,11 +559,45 @@ public boolean canRestore()
   @Override
   public void stopGracefully()
   {
-    log.info("Stopping gracefully.");
+    log.info("Stopping gracefully (status: [%s])", status);
     stopRequested = true;
-    if (runThread.isAlive()) {
-      log.info("Interrupting run thread (status: [%s])", status);
+
+    if (status == Status.PUBLISHING) {
       runThread.interrupt();
+      return;
+    }
+
+    try {
+      if (pauseLock.tryLock(15, TimeUnit.SECONDS)) {
+        try {
+          if (pauseRequested) {
+            pauseRequested = false;
+            shouldResume.signalAll();
+          }
+        }
+        finally {
+          pauseLock.unlock();
+        }
+      } else {
+        log.warn("While stopping: failed to acquire pauseLock before timeout, interrupting run thread");
+        runThread.interrupt();
+        return;
+      }
+
+      if (pollRetryLock.tryLock(15, TimeUnit.SECONDS)) {
```

Comment:
I'm not sure what the conventions are here, but would it make sense to extract the 15 seconds from here and the `pauseLock.tryLock` above, and perhaps de-hardcode?